packages/eslint-plugin/src/rules/no-partial-destructuring.ts

```diff
@@ -0,0 +1,447 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import {
+  createRule,
+  getConstrainedTypeAtLocation,
+  getParserServices,
+  getStaticMemberAccessValue,
+  isTypeFlagSet,
+} from '../util';
+
+interface PropertyDestructure {
+  /**
+   * @remarks We avoid computing property types until we know we need them.
+   */
+  type?: ts.Type;
+  property: TSESTree.Property;
+}
+
+type RemainingProperties = Map<number | string | symbol, PropertyDestructure>;
+
+type DynamicProperties = Set<PropertyDestructure>;
+
+export default createRule({
+  name: 'no-partial-destructuring',
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'Disallow destructuring patterns that do not fully handle all elements of inline object and tuple types',
+      requiresTypeChecking: true,
+    },
+    hasSuggestions: true,
+    messages: {
+      partialDestructuring: "Unused {{type}} '{{key}}' in destructuring.",
+      removeUnusedKey: "Remove unused {{type}} '{{key}}' from destructuring.",
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+    const compilerOptions = services.program.getCompilerOptions();
+    const isNoUncheckedIndexedAccess = compilerOptions.noUncheckedIndexedAccess;
+
+    /**
+     * Checks and reports a destructuring node and and its matching type annotation.
+     *
+     * _Ignores non-destructuring nodes._
+     */
+    function checkParam(param: TSESTree.Node, paramType: TSESTree.Node): void {
+      if (
+        param.type === AST_NODE_TYPES.ObjectPattern &&
+        paramType.type === AST_NODE_TYPES.TSTypeLiteral
+      ) {
+        return checkObject(param, paramType);
+      }
+
+      if (
+        param.type === AST_NODE_TYPES.ArrayPattern &&
+        paramType.type === AST_NODE_TYPES.TSTupleType
+      ) {
+        return checkTuple(param, paramType);
+      }
+
+      // do nothing otherwise
+    }
+
+    function checkObject(
+      node: TSESTree.ObjectPattern,
+      typeAnnotation: TSESTree.TSTypeLiteral,
+    ): void {
+      const remainingProperties: RemainingProperties = new Map();
+      const dynamicProperties: DynamicProperties = new Set();
+
+      // collect used properties
+      for (const property of node.properties) {
+        // bail on a rest element
+        if (property.type === AST_NODE_TYPES.RestElement) {
+          return;
+        }
+
+        const valueType = services.getTypeAtLocation(property.value);
+
+        // bail if misses; this is a type-error
```

Comment:
[Docs] I got confused by this and had to play with the test case.

```suggestion
        // bail if the property doesn't have a known type (that's a type error)
```